fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java

```diff
@@ -0,0 +1,2042 @@
+                .and(paramAlias, "CODE_SYSTEM_ID").eq(getCodeSystemId(targetResourceType))
+                )
+            .innerJoin(targetLR, alias(lrAlias),
+                on(lrAlias, "LOGICAL_ID").eq(paramAlias, "TOKEN_VALUE")
+                .and().coalesce(col(paramAlias, "REF_VERSION_ID"), col(lrAlias, "VERSION_ID")).eq(lrAlias, "VERSION_ID")
+                .and(lrAlias, "LOGICAL_RESOURCE_ID").inLiteralLong(logicalResourceIds)
+                .and(lrAlias, "IS_DELETED").eq().literal("N")
+                )
+            .innerJoin(parentR, alias(rAlias),
+                on(parentLRAlias, "CURRENT_RESOURCE_ID").eq(rAlias, "RESOURCE_ID")
+                )
+            ;
+
+        return queryData;
+    }
+
+    @Override
+    public QueryData addTokenParam(QueryData queryData, String resourceType, QueryParameter queryParm) throws FHIRPersistenceException {
+        // Add a join to the query. The NOT/NOT_IN modifiers are tricker because
+        // they need to be handled as a NOT EXISTS clause.
+        final int aliasIndex = getNextAliasIndex();
+        final SelectAdapter query = queryData.getQuery();
+        final Operator operator = getOperator(queryParm);
+        final String paramAlias = "P" + aliasIndex;
+        final String lrAlias = queryData.getLRAlias(); // join to LR at the same query level
+        final ExpNode filter;
+        if (operator == Operator.EQ) {
+            filter = getTokenFilter(queryParm, paramAlias).getExpression();
+        } else {
+            filter = getComplexTokenFilter(queryParm, paramAlias).getExpression();
+        }
+        // which table we join against depends on the fields used by the filter expression
+        final String xxTokenValues = getTokenParamTable(filter, resourceType, paramAlias);
+
+        String parameterName = queryParm.getCode();
+        // Append the suffix for :text modifier
+        if (Modifier.TEXT.equals(queryParm.getModifier())) {
+            parameterName += SearchConstants.TEXT_MODIFIER_SUFFIX;
+        }
+
+        if (queryParm.getModifier() == Modifier.NOT || queryParm.getModifier() == Modifier.NOT_IN) {
+            // Use a nested NOT EXISTS (...) instead of a simple join
+            SelectAdapter exists = Select.select("1");
+
+            if (operator == Operator.EQ) {
+                exists.from(xxTokenValues, alias(paramAlias))
+                .where(paramAlias, "LOGICAL_RESOURCE_ID").eq(lrAlias, "LOGICAL_RESOURCE_ID") // correlate with the main query
+                .and(paramAlias, "PARAMETER_NAME_ID").eq(getParameterNameId(parameterName));
+            } else {
+                exists.from(xxTokenValues, alias(paramAlias))
+                .where(paramAlias, "LOGICAL_RESOURCE_ID").eq(lrAlias, "LOGICAL_RESOURCE_ID") // correlate with the main query
+                .and(paramAlias, "PARAMETER_NAME_ID").eq(getParameterNameId(parameterName));
+            }
```

Comment:
not seeing any difference in the IF and ELSE blocks

Reply:
Good catch. They used to be different, but then I implemented the COMMON_TOKEN_VALUE_ID optimization by inspecting the filter expression...so I can eliminate this branch.